### PR TITLE
fix: upgrades `@walletconnect/ethereum-provider` and `@walletconnect/modal`

### PR DESCRIPTION
**Pull request by JFrankfurt**
These packages updates should remove polyfills, fix crypto.decode errors, allow empty required chains, and improve deep link ux.

Unfortunately, `@walletconnect/ethereum-provider` has some issues with exported types here. I've added a couple types and a util to guard for one of them. `ArrayOneOrMore<number>` must be enforced on either `chains` or `optionalChains` prior to the init call. This leads to a couple slightly awkward looking lines of code. 

```
export type ArrayOneOrMore<T> = {
  0: T
} & Array<T>

/**
 * This is a type guard for ArrayOneOrMore
 */
export function isArrayOneOrMore<T>(input: T[] = []): input is ArrayOneOrMore<T> {
  return input.length > 0
}
```

https://uniswapteam.slack.com/archives/C03R5G8T8BH/p1690896081926629?thread_ts=1690481148.038409&cid=C03R5G8T8BH

### CODE REVIEW

**Review by just-toby — DISMISSED** [review]

File: example/components/ConnectWithSelect.tsx [hunk 1]
```diff
@@ -26,14 +26,12 @@ function ChainSelect({
...
       </option>
-      <option value={-1} selected={activeChainId === -1}>
-        Default
-      </option>
+      <option value={-1}>Default</option>
       {chainIds.map((chainId) => (
-        <option key={chainId} value={chainId} selected={chainId === activeChainId}>
+        <option key={chainId} value={chainId}>
           {CHAINS[chainId]?.name ?? chainId}
         </option>
```
Comment (JFrankfurt) [inline example/components/ConnectWithSelect.tsx:36]: React was complaining about this invalid syntax in the example, but it's not related to the core diff here.

File: packages/walletconnect-v2/src/index.spec.ts [hunk 2]
```diff
@@ -68,7 +72,7 @@ class MockWalletConnectProvider extends MockEIP1193Provider<number> {
    * We mock this method later in the test suite to test behavior when optional chains are not supported.
    */
   public getConnectedChains() {
-    return this.opts.chains.concat(this.opts.optionalChains || [])
+    return (this.opts.chains || []).concat(this.opts.optionalChains || [])
```
Comment (JFrankfurt) [inline packages/walletconnect-v2/src/index.spec.ts:75]: since chains can be undefined now we need a default value

File: packages/walletconnect-v2/src/index.ts [hunk 3]
```diff
@@ -86,27 +102,54 @@ export class WalletConnect extends Connector {
     this.events.emit(URI_AVAILABLE, uri)
   }
 
+  private async initializeProvider(
```
Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:105]: I am wondering if this PR could be broken down into a smaller diff - the one that actually fixes the issues in PR title. I see a lot of changes like this one - which seem to be more of "refactoring" that are not related to the described issues, unless I am mistaken?

Would definitely understand the reason behind creating this utility function or reordering private class properties.

Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts:105]: I am also not aware of recent changes in other files, but one thing to keep in mind is that, in principle, we always tried to keep providers similar as far as top-level methods and private APIs they have. If you would look at other existing providers, you would notice they all have `isomorphicInitialize`, but not necessarily have a lot of private/non-standard methods. Something to keep in mind for future readability/ease of maintenance.

File: packages/walletconnect-v2/src/index.ts [hunk 4]
```diff
@@ -86,27 +102,54 @@ export class WalletConnect extends Connector {
...
+    return getChainsWithDefault(chains, desiredChainId)
+  }
+
+  private getChainProps(desiredChainId: number | undefined = this.defaultChainId): ChainsProps {
+    // Reorder chains and optionalChains if necessary
+    const orderedChains = this.reorderChainsBasedOnId(this.options.chains, desiredChainId)
+    const orderedOptionalChains = this.reorderChainsBasedOnId(this.options.optionalChains, desiredChainId)
+
+    // Validate and return the result
+    if (isArrayOneOrMore(orderedChains)) {
```
Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:138]: Why is this function necessary? Can't we just do `orderedChains.length > 0`?

File: packages/walletconnect-v2/src/index.ts [hunk 5]
```diff
@@ -86,27 +102,54 @@ export class WalletConnect extends Connector {
...
+      return { chains: orderedChains, optionalChains: orderedOptionalChains }
+    } else if (isArrayOneOrMore(orderedOptionalChains)) {
```
Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:140]: What is the difference between this branch of code and previous branch of code? Looks like you're literally executing same code. Could this be (if necessary) inlined into a single conditional, such as `if (isArrayOneOrMore(orderedChains) || isArrayOneOrMore(orderedOptionalChains)`?

File: packages/walletconnect-v2/src/index.ts [hunk 6]
```diff
@@ -48,27 +61,30 @@ export class WalletConnect extends Connector {
...
+  private readonly timeout: number
 
-  constructor({ actions, options, defaultChainId, timeout = DEFAULT_TIMEOUT, onError }: WalletConnectConstructorArgs) {
+  constructor({ actions, defaultChainId, options, timeout = DEFAULT_TIMEOUT, onError }: WalletConnectConstructorArgs) {
     super(actions, onError)
 
-    const { rpcMap, rpc, chains, ...rest } = options
-
-    this.options = rest
-    this.chains = chains
```
Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:66]: IMO this should not be removed. The principle was that all the props that are custom modified / not passed straight to WalletConnect should be kept outside to avoid confusion.

Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:66]: Also, I believe that, once this line of code is back, you should put the if statement here to check for number of chains (and optional chains) to ensure initialisation is valid and throw an error during the runtime.

Note that this can't be changed once created, so doing if condition here makes definitely more sense instead of at the lazy init time.

**Review by grabbou — CHANGES_REQUESTED** [review]

File: packages/walletconnect-v2/src/index.ts (hunk 6 above)
Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts:66]: I would also recommend putting `optionalChains` outside so that we can perform necessary reordering/check for presence.

File: packages/walletconnect-v2/src/index.ts [hunk 7]
```diff
@@ -86,27 +102,54 @@ export class WalletConnect extends Connector {
...
+
+  // Helper function to reorder chains array based on desiredChainId
+  private reorderChainsBasedOnId(
+    chains: number[] | undefined,
+    desiredChainId: number | undefined
+  ): number[] | undefined {
+    if (!chains || !desiredChainId || !chains.includes(desiredChainId) || chains.length === 0) {
+      return chains
+    }
+    return getChainsWithDefault(chains, desiredChainId)
```
Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:129]: This seems unnecessary to me. Why not include that `if` additional logic from lines 126-128 inside `getChainsWithDefault`? We spent a lot of time recently with @cmcewen thinking about the ways to name this function properly and while we debated `reoderChains`, we figured `getWithDefault` makes more sense to emphasise it's literally putting the first one as default.

**Review by grabbou — CHANGES_REQUESTED** [review]

File: packages/walletconnect-v2/src/index.ts [hunk 8]
```diff
@@ -86,27 +102,54 @@ export class WalletConnect extends Connector {
...
+
+    return this.handleProviderEvents(this.provider)
+  }
+
+  // Helper function to reorder chains array based on desiredChainId
+  private reorderChainsBasedOnId(
+    chains: number[] | undefined,
+    desiredChainId: number | undefined
+  ): number[] | undefined {
+    if (!chains || !desiredChainId || !chains.includes(desiredChainId) || chains.length === 0) {
```
Comment (grabbou) [inline packages/walletconnect-v2/src/index.ts:126]: @JFrankfurt 

This specifically

>  !chains.includes(desiredChainId) 

is what is causing the tests to fail. 

You should not handle this case here. If you take a look at `getChainsWithDefault` code, you will see it handles that case on its own:

https://github.com/Uniswap/web3-react/blob/505e164e156854e7a82bc56520ae0fab3129746d/packages/walletconnect-v2/src/utils.ts#L88-L94

**Review by grabbou — CHANGES_REQUESTED** [review]

File: packages/walletconnect-v2/src/index.ts (hunk 3 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:105]: Upgrading the package should (according to the wc team) get us the improvements mentioned in the title/description. All changes here are made to accommodate the new requirements from `@walletconnect/ethereum-provider`. i.e., enforcing `ArrayOfOneOrMore<T>` for `ChainProps`

File: packages/walletconnect-v2/src/index.ts (hunk 4 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:138]: unfortunately no. This is a type guard that explicitly returns the named type that WC requires. If we did `orderedChains.length > 0` we would still get `number[]` which TS interprets as a variable length array and therefore not compliant with the WC requirement that at least one of the arrays is length > 0

File: packages/walletconnect-v2/src/index.ts (hunk 7 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:129]: It's just a question of where we put the new type checks and changes. `getChainsWithDefault` does not support `chains` or `optionalChains` as arguments any longer with the new upstream `options` type changes from `@walletconnect/ethereum-provider`. We either need to diff the caller or diff the util. I chose the caller, but I don't have a preference between the two.

File: packages/walletconnect-v2/src/index.ts (hunk 6 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:66]: I was taking the opposite approach. If this is intended, we should at least include a comment explaining that we're intentionally *not* trying to be a thin pass-through layer.

Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:66]: this makes sense given the approach you outlined above

File: packages/walletconnect-v2/src/index.ts (hunk 5 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:140]: unfortunately this is not compatible with typescript's type discrimination system.

```
    if (isArrayOneOrMore(orderedChains) || isArrayOneOrMore(orderedOptionalChains)) {
      return { chains: orderedChains, optionalChains: orderedOptionalChains }
    }
```
gives the pictured error

image

File: example/pages/index.tsx [hunk 9]
```diff
@@ -13,7 +12,6 @@ export default function Home() {
       <div style={{ display: 'flex', flexFlow: 'wrap', fontFamily: 'sans-serif' }}>
         <MetaMaskCard />
         <WalletConnectV2Card />
-        <WalletConnectCard />
```
Comment (JFrankfurt) [inline example/pages/index.tsx:16]: removing broken example

File: packages/walletconnect-v2/src/index.ts (hunk 3 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:112]: @grabbou Are you saying here that you'd rather a class import a utility function from another file rather than define its own private utility method?

File: packages/walletconnect-v2/src/index.ts [hunk 10]
```diff
@@ -86,27 +111,47 @@ export class WalletConnect extends Connector {
...
+  ): ChainsProps {
+    // Reorder chains and optionalChains if necessary
+    const orderedChains = getChainsWithDefault(chains, desiredChainId)
+    const orderedOptionalChains = getChainsWithDefault(optionalChains, desiredChainId)
+
+    // Validate and return the result
+    if (isArrayOneOrMore(orderedChains)) {
+      return { chains: orderedChains, optionalChains: orderedOptionalChains }
+    } else if (isArrayOneOrMore(orderedOptionalChains)) {
+      return { optionalChains: orderedOptionalChains, chains: orderedChains }
```
Comment (zzmp) [inline packages/walletconnect-v2/src/index.ts:143]: return the natural ordering
```suggestion
      return { chains: orderedChains, optionalChains: orderedOptionalChains }
```

File: packages/walletconnect-v2/src/index.ts [hunk 11]
```diff
@@ -86,27 +111,47 @@ export class WalletConnect extends Connector {
...
+  private getChainProps(
+    chains: number[] | ArrayOneOrMore<number> | undefined,
+    optionalChains: number[] | ArrayOneOrMore<number> | undefined,
+    desiredChainId: number | undefined = this.defaultChainId
+  ): ChainsProps {
+    // Reorder chains and optionalChains if necessary
+    const orderedChains = getChainsWithDefault(chains, desiredChainId)
+    const orderedOptionalChains = getChainsWithDefault(optionalChains, desiredChainId)
+
+    // Validate and return the result
```
Comment (zzmp) [inline packages/walletconnect-v2/src/index.ts:139]:
```suggestion
    // Validate and return the result.
    // Type discrimination requires that we use these typeguard checks to guarantee a valid return type.
```

File: packages/walletconnect-v2/src/index.ts [hunk 12]
```diff
@@ -86,27 +111,47 @@ export class WalletConnect extends Connector {
...
-      }))
-
-      return provider
-        .on('disconnect', this.disconnectListener)
-        .on('chainChanged', this.chainChangedListener)
-        .on('accountsChanged', this.accountsChangedListener)
-        .on('display_uri', this.URIListener)
-    }))
+    this.eagerConnection = this.initializeProvider(desiredChainId)
+    return this.eagerConnection
```
Comment (zzmp) [inline packages/walletconnect-v2/src/index.ts:154]: maintain the existing codestyle
```suggestion
    return (this.eagerConnection = this.initializeProvider(desiredChainId))
```

File: packages/walletconnect-v2/src/index.ts [hunk 13]
```diff
@@ -86,27 +111,47 @@ export class WalletConnect extends Connector {
...
+    const rpcMap = this.rpcMap ? await getBestUrlMap(this.rpcMap, this.timeout) : undefined
+    const chainProps = this.getChainProps(this.chains, this.optionalChains, desiredChainId)
+
+    this.provider = await ethProviderModule.default.init({
+      ...this.options,
+      ...chainProps,
+      rpcMap,
+    })
+
+    return this.handleProviderEvents(this.provider)
```
Comment (zzmp) [inline packages/walletconnect-v2/src/index.ts:127]: You've introduced a waterfall pattern by modifying the rpcMap initialization:

```suggestion
    const rpcMap = this.rpcMap ? getBestUrlMap(this.rpcMap, this.timeout) : undefined
    const chainProps = this.getChainProps(this.chains, this.optionalChains, desiredChainId)

    this.provider = (await import('@walletconnect/ethereum-provider')).default.init({
      ...this.options,
      ...chainProps,
      rpcMap: await rpcMap,
    })

    return this.handleProviderEvents(this.provider)
```

Note that with this change, `rpcMap` can resolve while the provider module loads.

File: packages/walletconnect-v2/src/index.ts [hunk 14]
```diff
@@ -51,22 +64,34 @@ export class WalletConnect extends Connector {
...
+    this.chains = chains
+    this.optionalChains = optionalChains
+  }
+
+  private handleProviderEvents(provider: WalletConnectProvider): WalletConnectProvider {
+    return provider
+      .on('disconnect', this.disconnectListener)
+      .on('chainChanged', this.chainChangedListener)
+      .on('accountsChanged', this.accountsChangedListener)
+      .on('display_uri', this.URIListener)
```
Comment (zzmp) [inline packages/walletconnect-v2/src/index.ts:94]: This is only done once. I'd rather it just be inlined into `initializeProvider`, but I won't make a stink about it.

**Review by zzmp — CHANGES_REQUESTED** [review]

File: packages/walletconnect-v2/src/index.ts (hunk 3 above)
Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts:114]: > l changes here are made to accommodate the new requirements from @walletconnect/ethereum-provider. i.e., enforcing ArrayOfOneOrMore<T> for ChainProps

> @grabbou Are you saying here that you'd rather a class import a utility function from another file rather than define its own private utility method?

What are the utility functions that are relevant to the PR apart from the array check? I was referring to the private methods defined on the class, that, to me, look like just extracted from their original places and still being referred only in those places.

I would like to get a full picture - why those changes were necessary and how they relate to the PR

File: packages/walletconnect-v2/src/index.ts (hunk 4 above)
Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts:140]: Makes sense!

File: packages/walletconnect-v2/src/index.ts (hunk 6 above)
Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts:66]: I think this is exactly what we're trying to do - to be a thin pass-through layer. It's just that internally we have made this decision to differentiate between properties that we transform vs the ones we pass-through to make it easier to maintain.

That way, you can e.g. check for arrays during initialisation time and put them on class as a property for further reference.

File: packages/walletconnect-v2/src/index.ts (hunk 7 above)
Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts]: Yeah, I understand your mixed feelings there. I think - personally - since the only utility of that `getChainsWithDefault` was to cope with WC "abstraction" of choosing the preferred chain, I think putting it all there would make sense.

Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts]: BTW, just one tiny consideration for the future work: we could consider (if we're putting chains and optional chains as a class property) to e.g. make their type "arrayoneormore | undefined" and do not set them if they're either not defined or empty in length. That way, I think typings would become much easier down the road.

File: packages/walletconnect-v2/src/index.ts (hunk 14 above)
Comment (grabbou) [reply packages/walletconnect-v2/src/index.ts:94]: Yup, this is somewhat related to my previous comment https://github.com/Uniswap/web3-react/pull/868#discussion_r1284651907 and also +1 here. We should definitely keep things inline unless required by this PR.

File: packages/walletconnect-v2/src/index.ts (hunk 12 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:155]: ew I hate this pattern so much

File: packages/walletconnect-v2/src/index.ts (hunk 13 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts]: This.provider give the following error w/ this pattern

```Type 'Promise<EthereumProvider>' is missing the following properties from type 'EthereumProvider': events, namespace, accounts, signer, and 35 more.ts(2740)```

File: packages/walletconnect-v2/src/index.ts [hunk 15]
```diff
@@ -23,6 +23,19 @@ export type WalletConnectOptions = Omit<Parameters<typeof WalletConnectProvider.
   rpc?: { [chainId: number]: string | string[] }
 }
 
+/**
+ * Necessary type to interface with @walletconnect/ethereum-provider@2.9.2 which is currently unexported
```
Comment (zzmp) [inline packages/walletconnect-v2/src/index.ts:27]: You should make an upstream issue/PR to have that exported.

**Review by zzmp — APPROVED** [review]

File: packages/walletconnect-v2/src/index.ts (hunk 15 above)
Comment (JFrankfurt) [reply packages/walletconnect-v2/src/index.ts:27]: Talked to them about it when I first started working on this PR and they said they would take care of exporting it. May be worth checking in again.

**Review by just-toby — APPROVED** [review]
